server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java

```diff
@@ -192,9 +197,41 @@ ClusterState execute(ClusterState currentState) {
             }
 
             if (skippedSettings.isEmpty() == false && openIndices.isEmpty() == false) {
-                throw new IllegalArgumentException(
-                    String.format(Locale.ROOT, "Can't update non dynamic settings [%s] for open indices %s", skippedSettings, openIndices)
-                );
+                if (request.reopenShards()) {
+                    // We have non-dynamic settings and open indices. We will unassign all of the shards in these indices so that the new
+                    // changed settings are applied when the shards are re-assigned.
+                    routingTableBuilder = RoutingTable.builder(
+                        allocationService.getShardRoutingRoleStrategy(),
+                        currentState.routingTable()
+                    );
+                    for (Index index : openIndices) {
```

Comment:
I wonder, should we avoid reopening any indices for whom the change would be a no-op? Or would that be too much extra complexity?

Reply:
I don't think we want to do it in the general case (for dynamic settings or for static settings where reopen is not passed) because it could slightly change behavior and/or performance. But I can do it for this particular case. It would make these behave slightly differently than other setting changes, but I think that seems fine.